cirq/google/calibration/workflow.py

```diff
@@ -11,21 +11,210 @@
+            characterization. Defaults to sqrt_iswap_gates_translator.
+        merge_sub_sets: Whether to merge moments that can be characterized at the same time
+            together.
+        initial: The characterization requests obtained by a previous scan of another circuit. This
+            might be used to find a minimal set of moments to characterize across many circuits.
+
+    Returns:
+        Tuple of:
+          - list of calibration requests,
+          - list of indices of the generated characterization requests for each moment in the
+            supplied circuit. If None occurs at certain position, it means that the related moment
+            does not require characterization.
+
+    Raises:
+        IncompatibleMomentError when circuit contains a moment with operations other than the
+        operations matched by gates_translator, or it mixes a single qubit and two qubit gates.
+    """
+
+    def append_if_missing(calibration: FloquetPhasedFSimCalibrationRequest) -> int:
+        if calibration.pairs not in pairs_map:
+            index = len(calibrations)
+            calibrations.append(calibration)
+            pairs_map[calibration.pairs] = index
+            return index
+        else:
+            return pairs_map[calibration.pairs]
+
+    def merge_into_calibrations(calibration: FloquetPhasedFSimCalibrationRequest) -> int:
+        new_pairs = set(calibration.pairs)
+        for index in pairs_map.values():
+            assert calibration.gate == calibrations[index].gate
+            assert calibration.options == calibrations[index].options
+            existing_pairs = calibrations[index].pairs
+            if new_pairs.issubset(existing_pairs):
+                return index
+            elif new_pairs.issuperset(existing_pairs):
+                calibrations[index] = calibration
+                return index
+            else:
+                new_qubit_pairs = calibration.qubit_pairs
+                existing_qubit_pairs = calibrations[index].qubit_pairs
+                if all(
+                    (
+                        new_qubit_pairs[q] == existing_qubit_pairs[q]
+                        for q in set(new_qubit_pairs.keys()).intersection(
+                            existing_qubit_pairs.keys()
+                        )
+                    )
+                ):
+                    calibrations[index] = FloquetPhasedFSimCalibrationRequest(
+                        gate=calibration.gate,
+                        pairs=tuple(sorted(new_pairs.union(existing_pairs))),
+                        options=options,
+                    )
+                    return index
+
+        index = len(calibrations)
+        calibrations.append(calibration)
+        pairs_map[calibration.pairs] = index
+        return index
+
+    if initial is None:
+        calibrations = []
+        moments_map = []
+    else:
+        calibrations, moments_map = initial
+
+    pairs_map = {}
```

Comment:
does this work? I'd put this above all the function definitions that use it if only for readability

Reply:
Works, done.